### PR TITLE
Function calling

### PR DESCRIPTION
**Pull request by fakhirali**

### CODE REVIEW

**Comment by fakhirali** [comment]
Implements function calling in gpt llm. You can now pass the tools, their functions and utterances that would be played when the function is called.

File: openvoicechat/llm/llm_gpt.py [hunk 1]
```diff
@@ -3,33 +3,102 @@
...
+            for chunk in stream:
+                finish_reason = chunk.choices[0].finish_reason
+                if chunk.choices[0].delta.tool_calls is not None:
+                    function_call_detected = True
+                    tool_call = chunk.choices[0].delta.tool_calls[0]
+                    if tool_call.function.name:
+                        func_call["name"] = tool_call.function.name
+                        func_call["id"] = tool_call.id
+                        func_call['arguments'] = ''
+                        yield random.choice(self.tool_utterances[func_call['name']]) + ' . ' # the period is to make
```
Comment (haiderrzaidi) [inline openvoicechat/llm/llm_gpt.py:67]: better to add comments to explain whats happening

File: openvoicechat/utils.py [hunk 2]
```diff
@@ -81,28 +83,31 @@ class Player_ws:
...
     def play(self, audio_array, samplerate):
+        self.playing = True
+        duration = len(audio_array) / samplerate
         if audio_array.dtype == np.int16:
             audio_array = audio_array / (1 << 15)
         audio_array = audio_array.astype(np.float32)
         audio_array = librosa.resample(y=audio_array, orig_sr=samplerate, target_sr=44100)
         audio_array = audio_array.tobytes()
         self.output_queue.put(audio_array)
+        if self._timer_thread is not None:
```
Comment (haiderrzaidi) [inline openvoicechat/utils.py:97]: i think adding comments would be better here to to explain the complex logic to the users.

**Review by haiderrzaidi — APPROVED** [review]
overall the changes made aren't very complex but better comments would be nice and i see the there is not much error handling done so i think its better to add that too

**Comment by fakhirali** [comment]
Will add comments, there is a some documentation missing as well, will make a different PR for it.

File: openvoicechat/llm/llm_gpt.py (hunk 1 above)
Comment (fakhirali) [reply openvoicechat/llm/llm_gpt.py:68]: Comment added

File: openvoicechat/utils.py (hunk 2 above)
Comment (fakhirali) [reply openvoicechat/utils.py:98]: Comment added

**Pull request merged**